### PR TITLE
feat: add additional functionality to the VM module - `avm/res/compute/virtual-machine`

### LINKED ISSUE #1345: [AVM Module Issue]: Create custom NIC without the need of nicSuffix
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

I'm not sure

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Is it possible to create the NIC name with a custom name? Without the need of the nicSuffix?

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1347: [AVM Module Issue]: Virtual Machine settings missing
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Dear AVM team, 

When working on AUMv2, there are two settings required when managing a Virtual Machine under it:

- AutomaticByPlatform
- bypassPlatformSafetyChecksOnUserSchedule

The last one is only missing in the windowsConfiguration variable (and Linux):

```
var windowsConfiguration = {
  provisionVMAgent: provisionVMAgent
  enableAutomaticUpdates: enableAutomaticUpdates
  patchSettings: (provisionVMAgent && (patchMode =~ 'AutomaticByPlatform' || patchMode =~ 'AutomaticByOS' || patchMode =~ 'Manual')) ? {
    patchMode: patchMode
    assessmentMode: patchAssessmentMode
  } : null
  timeZone: empty(timeZone) ? null : timeZone
  additionalUnattendContent: empty(additionalUnattendContent) ? null : additionalUnattendContent
  winRM: !empty(winRM) ? {
    listeners: winRM
  } : null
}
```

Is it possible to apply the following update as such:

```
@description('Optional. Enables customer to schedule patching without accidental upgrades.')
param bypassPlatformSafetyChecksOnUserSchedule bool = true

@description('Optional. Specifies the reboot setting for all AutomaticByPlatform patch installation operations.')
@allowed([
  'Always'
  'IfRequired'
  'Never'
  'Unknown'
])
param rebootSetting string = 'IfRequired'

var windowsConfiguration = {
  provisionVMAgent: provisionVMAgent
  enableAutomaticUpdates: enableAutomaticUpdates
  patchSettings: (provisionVMAgent && (patchMode =~ 'AutomaticByPlatform' || patchMode =~ 'AutomaticByOS' || patchMode =~ 'Manual'))
    ? {
        patchMode: patchMode
        assessmentMode: patchAssessmentMode
        automaticByPlatformSettings: {
          bypassPlatformSafetyChecksOnUserSchedule: bypassPlatformSafetyChecksOnUserSchedule
          rebootSetting: rebootSetting
        }
      }
    : null
  timeZone: empty(timeZone) ? null : timeZone
  additionalUnattendContent: empty(additionalUnattendContent) ? null : additionalUnattendContent
  winRM: !empty(winRM)
    ? {
        listeners: winRM
      }
    : null
}
```

Would love to hear from you!

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1437: [AVM Module Issue]: Enable Auto-Shutdown in VM module
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Is it possible to add the auto-shutdown option in the VM module?

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1446: [AVM Module Issue]: extensionDependencyAgentConfig in virtual-machine module missing enableAMA in Settings
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

0.2.2

### Description

The VM Extension Dependency Agent is missing the enableAMA. Without this the DependencyAgent sends its data to the Legacy LAWW agent instead of the AMA agent. ref: https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/agent-dependency-windows#extension-schema

example ARM template for DepencyAgent where 'enableAMA' is set to True:
``` arm
    {
      "type": "Microsoft.Compute/virtualMachines/extensions",
      "name": "[concat(parameters('vmName'),'/DAExtension')]",
      "apiVersion": "[variables('vmExtensionsApiVersion')]",
      "location": "[resourceGroup().location]",
      "dependsOn": [],
      "properties": {
          "publisher": "Microsoft.Azure.Monitoring.DependencyAgent",
          "type": "DependencyAgentWindows",
          "typeHandlerVersion": "9.10",
          "autoUpgradeMinorVersion": true,
          "settings": {
                "enableAMA": "true"
		    }
      }
```

Suggesting to add a 'enableAMA' property to 'extensionDependencyAgentConfig'. example:

```bicep
      extensionDependencyAgentConfig: {
        enabled: true
        enableAMA: true
      }
```

gr, pb

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1496: [AVM Module Issue]: VM `availabilityZone` parameter should be renamed to `zone`
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

In alignment with the latest discussions, the `zones` implementation for the VM is mostly aligned, but should be changed from `availabilityZone` to `zone` - which we'll soon specify in the AVM specs too.

Reference: https://github.com/Azure/Azure-Verified-Modules/pull/805

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1516: [AVM Module Issue]:  enableIPForwarding doesn't accept true when deploying NIC through virtual-machine
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Describe the bug
If i try to deploy a new virtual machine and when i set enableIPForwarding to true, i have a message arror

To reproduce
create a vm with enableipforwarding set to true

Code snippet
param enableIPForwarding bool = true
var nicConfigurations = [
    {
      nicSuffix: '-nic-01'
      enableIPForwarding: enableIPForwarding
      ipConfigurations: [
        {
          name: 'ipconfig1'
          subnetResourceId: subnetResourceId
          privateIPAllocationMethod: IPallocation
          privateIPAddress: privateIPAddress
          primary: true
        }
      ]
    }
]
Relevant log output
New-AzResourceGroupDeployment: 3:28:20 PM - The deployment 'main' failed with error(s). Showing 2 out of 2 error(s).
Status Message: Unable to process template language expressions for resource '/subscriptions/************/resourceGroups/***/providers/Microsoft.Resources/deployments/rusaxk4qxrx5o-VM-Nic-0' at line '1' and column '29126'. 'The template function 'empty' expects its parameter to be an object, an array, or a string. The provided value is of type 'Boolean'. Please see https://aka.ms/arm-function-empty for usage details.' (Code:InvalidTemplate)
Status Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details. (Code: DeploymentFailed)
 - Unable to process template language expressions for resource '/subscriptions/**************/resourceGroups/*******/providers/Microsoft.Resources/deployments/rusaxk4qxrx5o-VM-Nic-0' at line '1' and column '29126'. 'The template function 'empty' expects its parameter to be an object, an array, or a string. The provided value is of type 'Boolean'. Please see https://aka.ms/arm-function-empty for usage details.' (Code:InvalidTemplate)
CorrelationId: ece39906-a82e-468e-8121-01ba7bd12b23

### (Optional) Correlation Id

_No response_

### LINKED ISSUE #1630: [AVM Module Issue]: Virtual Machine has no support for Custom Automanage configuration profiles
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Version

0.2.3

### Description

The Virtual-Machine module does not support custom Automanage profiles. It only allows you to use the default ones.

When trying to deploy a VM with a custom Automanage profile: Microsoft.Automanage/configurationProfiles@2022-05-04 it is not allowed per module main.bicep restriction on the parameter:
_The property "configurationProfile" expected a value of type "'' | '/providers/Microsoft.Automanage/bestPractices/AzureBestPracticesDevTest' | '/providers/Microsoft.Automanage/bestPractices/AzureBestPracticesProduction' | null" but the provided value is of type "string".bicep(BCP036)_

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by rahalan**
run (failed because of quotas and wrong tenant specific SPN guid): [avm.res.compute.virtual-machine](https://github.com/rahalan/bicep-registry-modules/actions/workflows/avm.res.compute.virtual-machine.yml)

added features:

- NIC name can now be completely customized, see #1345 
- rename availabilityZone to zone, see #1496 
- adding AMA setting to dependency agent, see #1446 
- adding auto-shutdown, see #1437 
- adding additional settings for Update Management, see #1347 
- fixing issue to enable IP forwarding, see #1516 
- fixing automanage custom profile, see #1630

### CODE REVIEW

**Comment by AlexanderSehr** [comment]
@ChrisSidebotham / @eriqua could you take a look at this one?

**Review by eriqua — COMMENTED** [review]
Hi @rahalan, sorry for keeping you waiting, just a couple of things meanwhile reviewing:
- Please resolve conflicts
- Do you have a link to a green pipeline, tested before changing the SPN value back to the brm one?
- Are necessary quotas available in the target brm testing subscription?

The reson for the last 2 points is we need to make sure the deployment validation succeeds. Otherwise the updated version of the module will not get published

**Comment by rahalan** [comment]
> Hi @rahalan, sorry for keeping you waiting, just a couple of things meanwhile reviewing:
> 
> * Please resolve conflicts
> * Do you have a link to a green pipeline, tested before changing the SPN value back to the brm one?
> * Are necessary quotas available in the target brm testing subscription?
> 
> The reson for the last 2 points is we need to make sure the deployment validation succeeds. Otherwise the updated version of the module will not get published

@eriqua conflicts are resolved now (was just a json hash). Both quotas and SPNs are already in main and working.

**Comment by eriqua** [comment]
> > Hi @rahalan, sorry for keeping you waiting, just a couple of things meanwhile reviewing:
> > 
> > * Please resolve conflicts
> > * Do you have a link to a green pipeline, tested before changing the SPN value back to the brm one?
> > * Are necessary quotas available in the target brm testing subscription?
> > 
> > The reson for the last 2 points is we need to make sure the deployment validation succeeds. Otherwise the updated version of the module will not get published
> 
> @eriqua conflicts are resolved now (was just a json hash). Both quotas and SPNs are already in main and working.

I'm sure about that. Still one successful run is needed in order to prove the updates work. Without it is the same as not having the updates tested. Do we have at least one successful run in your fork related to this change?

**Comment by rahalan** [comment]
@eriqua please look at the last run from my fork (link is on top of the issue). Two tests failed: nVidia one because of MCAPS policy restriction and the Linux max one (which was earlier already green) because (again) of a KeyVault name already being used.

File: avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep:223]: Placeholder to make sure we change this back before merge as a last change

File: avm/res/compute/virtual-machine/tests/e2e/waf-aligned/dependencies.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/tests/e2e/waf-aligned/dependencies.bicep:223]: Same here

Comment (eriqua) [reply avm/res/compute/virtual-machine/tests/e2e/waf-aligned/dependencies.bicep:223]: Closing as likely correct already

File: avm/res/compute/virtual-machine/main.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/main.bicep:343]: Is this relevant for linux too? https://learn.microsoft.com/en-us/azure/templates/microsoft.compute/virtualmachines?pivots=deployment-language-terraform#linuxvmguestpatchautomaticbyplatformsettings-2

File: avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/windows.max/dependencies.bicep:223]: @eriqua  That is the correct one

File: avm/res/compute/virtual-machine/main.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/main.bicep:570]: I'd suggest to keep `name` instead of `customNicName` for the `nicConfiguration` property. Also, can we add this option to the max test?

File: avm/res/compute/virtual-machine/main.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:318]: Done

File: avm/res/compute/virtual-machine/main.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep]: Done

File: avm/res/compute/virtual-machine/main.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep]: Done

File: avm/res/compute/virtual-machine/main.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/main.bicep:625]: Are we sure about all these default values?

File: avm/res/compute/virtual-machine/main.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/main.bicep:124]: It would be ideal to define complex objects like this with user defined types.
I'll not block the PR on that, but just for awareness. @rahalan you can resolve once read.

File: avm/res/compute/virtual-machine/main.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:625]: @eriqua  That is what you get in the portal experience

File: avm/res/compute/virtual-machine/main.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:124]: @eriqua I know, will do in a later iteration

File: avm/res/compute/virtual-machine/main.bicep
Comment (eriqua) [reply avm/res/compute/virtual-machine/main.bicep:625]: Fair. When implementing user defined types we'll be able to properly document those defaults

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**
